File: backend/src/services/auth/identity.rs
use crate::models::Account;
use crate::repositories::{AccountRepository, ReaderRepository};
use crate::utils::jwt::generate_jwt;
use bson::oid::ObjectId;
use mongodb::Database;

pub struct OAuthUserPayload {
    pub provider: String,
    pub provider_id: String,
    pub name: String,
    pub email: Option<String>,
    pub avatar: Option<String>,
    pub handle: Option<String>,
    pub access_token: String,
    pub scope: Option<String>,
}

pub struct IdentityService {
    db: Database,
    jwt_secret: String,
}

impl IdentityService {
    pub fn new(db: &Database, jwt_secret: String) -> Self {
        Self {
            db: db.clone(),
            jwt_secret,
        }
    }

    /// 处理通用的 OAuth 登录流逻辑
    pub async fn process_oauth_login(
        &self,
        payload: OAuthUserPayload,
    ) -> Result<(ObjectId, bool, bool), String> {
        let reader_repo = ReaderRepository::new(&self.db);
        let account_repo = AccountRepository::new(&self.db);

        log::info!(
            "[OAuth] 开始处理登录: provider={}, provider_id={}, name={}, email={:?}",
            payload.provider,
            payload.provider_id,
            payload.name,
            payload.email
        );

        // 1. 查找现有账号
        let existing_acc = account_repo
            .find_by_provider_and_account_id(&payload.provider, &payload.provider_id)
            .await
            .map_err(|e| e.to_string())?;

        if let Some(acc) = existing_acc {
            log::info!(
                "[OAuth] 找到现有账号: account_id={}, user_id={}",
                acc.id,
                acc.user_id
            );
            let reader = reader_repo
                .find_by_id(acc.user_id)
                .await
                .map_err(|e| e.to_string())?
                .ok_or_else(|| "账户数据不一致".to_string())?;
            log::info!("[OAuth] 返回现有用户: is_new_user=false");
            return Ok((reader.id, reader.is_owner, false));
        }

        log::info!("[OAuth] 未找到现有账号，尝试邮箱匹配");

        // 2. 新用户：尝试通过邮箱自动匹配
        if let Some(ref email) = payload.email {
            if let Some(matched_reader) = reader_repo
                .find_by_email(email)
                .await
                .map_err(|e| e.to_string())?
            {
                log::info!(
                    "[OAuth] 通过邮箱匹配到现有用户: email={}, user_id={}",
                    email,
                    matched_reader.id
                );
                // 自动绑定到已有 Reader
                self.create_account_record(matched_reader.id, &payload)
                    .await?;
                log::info!("[OAuth] 自动绑定完成: is_new_user=false");
                return Ok((matched_reader.id, matched_reader.is_owner, false));
            }
        }

        log::info!("[OAuth] 创建新用户");

        // 3. 彻底的新用户：检查是否是系统第一个用户
        let is_first = reader_repo.is_empty().await.unwrap_or(false);
        let temp_id = ObjectId::new();

        self.create_account_record(temp_id, &payload).await?;

        log::info!(
            "[OAuth] 新用户创建完成: temp_id={temp_id}, is_first={is_first}, is_new_user=true"
        );

        Ok((temp_id, is_first, true))
    }

    /// 辅助函数：内部创建 Repository，避开生命周期烦恼
    async fn create_account_record(
        &self,
        user_id: ObjectId,
        payload: &OAuthUserPayload,
    ) -> Result<(), String> {
        let account_repo = AccountRepository::new(&self.db);

        let mut account = if payload.provider == "github" {
            let github_id = payload
                .provider_id
                .parse::<u64>()
                .map_err(|_| "无效的 GitHub ID")?;
            Account::new_github_with_info(
                user_id,
                github_id,
                payload.access_token.clone(),
                payload.scope.clone(),
                payload.name.clone(),
                payload.email.clone(),
                payload.avatar.clone().unwrap_or_default(),
                payload.handle.clone().unwrap_or_else(|| "user".to_string()),
            )
        } else {
            Account::new_qq_with_info(
                user_id,
                payload.provider_id.clone(),
                payload.access_token.clone(),
                payload.name.clone(),
                payload.avatar.clone().unwrap_or_default(),
            )
        };

        // 如果是 QQ 或其他，确保 provider 字段正确
        if payload.provider != "github" {
            account.provider = payload.provider.clone();
        }

        account_repo
            .create_account(&account)
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn merge_identities(&self, from_id: ObjectId, to_id: ObjectId) -> Result<(), String> {
        let account_repo = AccountRepository::new(&self.db);
        let accounts = account_repo
            .find_by_user_id(from_id)
            .await
            .map_err(|e| e.to_string())?;
        for acc in accounts {
            account_repo
                .update_user_id(acc.id, to_id)
                .await
                .map_err(|e| e.to_string())?;
        }
        Ok(())
    }

    pub fn issue_token(&self, user_id: ObjectId, is_owner: bool) -> Result<String, String> {
        generate_jwt(user_id, is_owner, &self.jwt_secret).map_err(|e| format!("{e:?}"))
    }
}

File: backend_ts/src/services/identity.ts
/**
 * 身份服务
 *
 * 处理 OAuth 登录、账号关联、临时账户等复杂的业务逻辑
 */

import type { OAuthUserInfo } from "./oauth";
import { Elysia } from "elysia";
import { ObjectId } from "mongodb";
import { getConfig } from "@/config";
import { db_count, db_insert, db_read, db_update } from "@/lib/db";
import { generateToken, jwtPlugin } from "@/middleware/auth";

const config = getConfig();
const DB_NAME = config.mongodb.database || "mx-space";

/**
 * OAuth 用户 Payload
 */
export interface OAuthUserPayload {
  provider: string;
  providerId: string;
  name: string;
  email?: string;
  avatar?: string;
  handle?: string;
  accessToken: string;
  scope?: string;
}

/**
 * 身份服务
 */
export class IdentityService {
  private jwtInstance: any;

  constructor() {
    // 创建 JWT 实例
    const jwtApp = new Elysia().use(jwtPlugin());
    this.jwtInstance = (jwtApp as any).decorator.jwt;
  }

  /**
   * 处理通用的 OAuth 登录流逻辑
   *
   * @param payload - OAuth 用户信息
   * @returns [userId, isOwner, isNewUser]
   */
  async processOAuthLogin(
    payload: OAuthUserPayload,
  ): Promise<[ObjectId, boolean, boolean]> {
    // 1. 查找现有账号
    const existingAccounts = await db_read(
      DB_NAME,
      "accounts",
      {
        provider: payload.provider,
        accountId: payload.providerId,
      },
      {},
    );

    if (existingAccounts.length > 0) {
      const account = existingAccounts[0];
      // 查找关联的 Reader
      const readers = await db_read(
        DB_NAME,
        "readers",
        { _id: new ObjectId(account.userId) },
        {},
      );

      if (readers.length > 0) {
        const reader = readers[0];
        return [new ObjectId(reader._id), reader.isOwner || false, false];
      }

      throw new Error("账户数据不一致");
    }

    // 2. 新用户：尝试通过邮箱自动匹配
    if (payload.email) {
      const matchedReaders = await db_read(
        DB_NAME,
        "readers",
        { email: payload.email },
        {},
      );

      if (matchedReaders.length > 0) {
        const matchedReader = matchedReaders[0];
        // 自动绑定到已有 Reader
        await this.createAccountRecord(new ObjectId(matchedReader._id), payload);
        return [new ObjectId(matchedReader._id), matchedReader.isOwner || false, false];
      }
    }

    // 3. 彻底的新用户：检查是否是系统第一个用户
    const readerCount = await db_count(DB_NAME, "readers", {});
    const isFirst = readerCount === 0;
    const tempId = new ObjectId();

    await this.createAccountRecord(tempId, payload);

    return [tempId, isFirst, true];
  }

  /**
   * 创建 Account 记录
   */
  private async createAccountRecord(
    userId: ObjectId,
    payload: OAuthUserPayload,
  ): Promise<void> {
    const now = new Date();

    const account: any = {
      userId,
      provider: payload.provider,
      accountId: payload.providerId,
      accessToken: payload.accessToken,
      scope: payload.scope || null,
      oauthName: payload.name,
      oauthEmail: payload.email || null,
      oauthAvatar: payload.avatar || "",
      oauthHandle: payload.handle || (payload.provider === "github" ? "user" : ""),
      createdAt: now,
      updatedAt: now,
    };

    await db_insert(DB_NAME, "accounts", account);
  }

  /**
   * 合并身份（将一个用户的所有账号转移到另一个用户）
   */
  async mergeIdentities(fromId: ObjectId, toId: ObjectId): Promise<void> {
    const accounts = await db_read(
      DB_NAME,
      "accounts",
      { userId: fromId },
      {},
    );

    for (const account of accounts) {
      await db_update(
        DB_NAME,
        "accounts",
        { _id: new ObjectId(account._id) },
        { userId: toId },
      );
    }
  }

  /**
   * 颁发 JWT token（与 Rust 后端一致）
   */
  async issueToken(userId: ObjectId, isOwner: boolean): Promise<string> {
    return generateToken(this.jwtInstance, userId.toHexString(), isOwner);
  }
}

/**
 * 将 OAuthUserInfo 转换为 OAuthUserPayload
 */
export function convertToPayload(info: OAuthUserInfo): OAuthUserPayload {
  return {
    provider: info.provider,
    providerId: info.providerId,
    name: info.nickname,
    email: info.email,
    avatar: info.avatar,
    handle: undefined,
    accessToken: info.accessToken || "",
    scope: undefined,
  };
}
